### PR TITLE
Support for Vector: VectorValue

### PR DESCRIPTION
**Pull request by zhengyangl**

### CODE REVIEW

File: ir/value.cpp [hunk 1]
```diff
@@ -110,4 +111,37 @@ expr Input::getTyVar() const {
   return expr::mkVar(tyname.c_str(), 2);
 }
 
+
+static string genVectorName(const std::vector<Value *> &vals) {
+  ostringstream ss;
+  ss << "<";
+  for (unsigned idx = 0; idx < vals.size(); ++idx) {
+    ss << *vals[idx];
+    if (idx != vals.size() -1)
```
Comment (nunoplopes) [inline ir/value.cpp]: it's easier to check for idx != 0 and pre-add a comma

File: ir/value.cpp [hunk 2]
```diff
@@ -110,4 +111,37 @@ expr Input::getTyVar() const {
   return expr::mkVar(tyname.c_str(), 2);
 }
 
+
+static string genVectorName(const std::vector<Value *> &vals) {
```
Comment (nunoplopes) [inline ir/value.cpp]: we don't use std:: in .cpp files

File: ir/value.cpp [hunk 3]
```diff
@@ -110,4 +111,37 @@ expr Input::getTyVar() const {
...
+  UNREACHABLE();
+}
+
+StateValue VectorValue::toSMT(State &s) const {
+  expr val;
+  expr not_poison = true;
+
+  for (unsigned idx = 0; idx < vals.size(); ++idx) {
+    auto [v, vs] = vals[idx]->toSMT(s);
+    val = idx == 0 ? v : val.concat(v);
```
Comment (nunoplopes) [inline ir/value.cpp]: float values need to be cast to bit-vectors.
I would suggest we introduce a new function in the VectorType that creates an SMT value from a vector of individual values (to avoid repeating this code to cast to BV, as it will probably needed later).

File: ir/value.cpp [hunk 4]
```diff
@@ -110,4 +111,37 @@ expr Input::getTyVar() const {
...
+    not_poison &= vs;
```
Comment (nunoplopes) [inline ir/value.cpp]: poison is per element, so we need concat here as well. I think the function just mention can take a vector of StateValues and return a StateValue

File: ir/value.cpp [hunk 5]
```diff
@@ -110,4 +111,37 @@ expr Input::getTyVar() const {
...
+  }
+
+  return { move(val), true };
```
Comment (nunoplopes) [inline ir/value.cpp]: true is certainly not correct here :)

**Comment by nunoplopes** [comment]
Today I had to do quite a few changes related with vectors (and aggregate types in general) in order to support checking of some miscompilation in LLVM we were chasing. In the process I did a large refactoring of the aggregate types (arrays/structs/vectors) to reduce code duplication.
Could you please have a look?  This may have clashed with this patch, I'm sorry.

**Comment by zhengyangl** [comment]
Sure, thank you @nunoplopes for the review and the notify of this potential clash. I will be a little occupied before the CGO deadline (this Friday). I will take care of all the reviews afterwards.

File: ir/value.cpp (hunk 1 above)
Comment (zhengyangl) [reply ir/value.cpp]: fixed :)

File: ir/value.cpp (hunk 2 above)
Comment (zhengyangl) [reply ir/value.cpp]: fixed :)

File: ir/value.cpp (hunk 4 above)
Comment (zhengyangl) [reply ir/value.cpp]: Hi @nunoplopes , I've fixed this and If you're okay with the code I'll send more code for review (parser and vector operations).

File: ir/value.cpp (hunk 5 above)
Comment (zhengyangl) [reply ir/value.cpp]: oops:)

File: ir/type.cpp [hunk 6]
```diff
@@ -664,6 +668,20 @@ VectorType::VectorType(string &&name, unsigned elements, Type &elementTy)
...
 
+StateValue VectorType::aggregateVals(const vector<StateValue> &vals) const {
+  assert(defined);
+  assert(vals.size() == elements);
+
+  expr val;
+  expr not_poison;
+  for (unsigned idx = 0; idx < elements; ++idx) {
+    auto [v, vs] = children[idx]->toBV(vals[idx]);
+    val = idx == 0 ? v : val.concat(v);
```
Comment (nunoplopes) [inline ir/type.cpp]: StateValue now supports concat. You can do this in one go :)
See AggregateConst::toSMT().  Actually that code should be replaced with a call to this.

Comment (nunoplopes) [reply ir/type.cpp]: BTW, I think this method can be hoisted to the AggregateType class. The extract is also there.

File: ir/value.h [hunk 7]
```diff
@@ -83,4 +83,12 @@ class Input final : public Value {
   smt::expr getTyVar() const;
 };
 
+class VectorValue final : public Value {
```
Comment (nunoplopes) [inline ir/value.h]: Is this class still needed?  I added a ConstantAggregate class that should be sufficient for vectors, structs and arrays.

File: ir/type.cpp (hunk 6 above)
Comment (zhengyangl) [reply ir/type.cpp]: Fixed, these makes code extremely clean!

File: ir/value.h (hunk 7 above)
Comment (zhengyangl) [reply ir/value.h]: Good point, I've removed this function.

**Comment by zhengyangl** [comment]
@nunoplopes This PR is ready for review again:) I'll send some real caller to these updates in the next pull request!

File: ir/type.h [hunk 8]
```diff
@@ -229,6 +229,7 @@ class AggregateType : public Type {
     smt::expr(*map)(const IR::StateValue&, const IR::StateValue&),
     smt::expr(*reduce)(const std::set<smt::expr>&),
     const IR::StateValue &a, const IR::StateValue &b) const override;
+  StateValue aggregateVals(const std::vector<StateValue> &vals) const;
```
Comment (nunoplopes) [inline ir/type.h]: seems like a tab here?

File: ir/type.cpp [hunk 9]
```diff
@@ -637,6 +637,18 @@ expr AggregateType::map_reduce(expr(*map)(const StateValue&, const StateValue&),
   return reduce(r);
 }
 
+StateValue AggregateType::aggregateVals(const vector<StateValue> &vals) const {
+  assert(defined);
```
Comment (nunoplopes) [inline ir/type.cpp]: this assert isn't valid. `defined` is used for types that are fully defined (nothing symbolic).
So if the type is inferred, `defined` will be false, though the type will be then fine for consumption.
TL;DR: just remove this assert :)

File: ir/type.h (hunk 8 above)
Comment (zhengyangl) [reply ir/type.h]: No my emacs would never produce evil tab :), see the above lines for the indentation.

File: ir/type.cpp (hunk 9 above)
Comment (zhengyangl) [reply ir/type.cpp]: fixed :)
